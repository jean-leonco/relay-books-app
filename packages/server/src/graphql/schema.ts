import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../core/middlewares/applyMiddlewares';

// import SubscriptionType from './subscription/SubscriptionType';
import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const _schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  // subscription: SubscriptionType,
});

applyMiddlewares(_schema);

export const schema = _schema;
