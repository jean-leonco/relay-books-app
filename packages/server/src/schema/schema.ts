import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../graphql/middlewares/applyMiddlewares';

import MutationType from './MutationType';
import QueryType from './QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

applyMiddlewares(schema);

export default schema;
