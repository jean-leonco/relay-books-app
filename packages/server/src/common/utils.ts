import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
