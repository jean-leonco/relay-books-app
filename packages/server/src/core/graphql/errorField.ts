import { GraphQLString } from 'graphql';

const errorField = {
  error: {
    type: GraphQLString,
    description: 'Default error field resolver for mutations',
    resolve: ({ error }) => error,
  },
};

export default errorField;
