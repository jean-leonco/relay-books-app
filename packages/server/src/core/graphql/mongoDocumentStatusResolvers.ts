import { GraphQLBoolean, GraphQLString } from 'graphql';

export const mongoDocumentStatusResolvers = {
  isActive: {
    type: GraphQLBoolean,
    description: 'The soft delete status.',
    resolve: (obj) => obj.isActive,
  },
  removedAt: {
    type: GraphQLString,
    description: 'The hard delete status.',
    resolve: (obj) => (obj.removedAt ? obj.removedAt.toISOString() : null),
  },
  createdAt: {
    type: GraphQLString,
    description: 'The date that the node was created.',
    resolve: (obj) => (obj.createdAt ? obj.createdAt.toISOString() : null),
  },
  updatedAt: {
    type: GraphQLString,
    description: 'The date that the node was last updated.',
    resolve: (obj) => (obj.updatedAt ? obj.updatedAt.toISOString() : null),
  },
};
