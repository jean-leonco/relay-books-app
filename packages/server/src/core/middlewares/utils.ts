import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
