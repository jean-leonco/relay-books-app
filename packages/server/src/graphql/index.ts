/* eslint-disable no-console */
import 'core-js/stable';
import { createServer } from 'http';

import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';

import { GRAPHQL_PORT, GRAPHQL_HOST } from '../common/config';
import connectDatabase from '../common/database';

import app from './app';
import { schema } from './schema';

type ConnectionParams = {
  authorization?: string;
};

const runServer = async () => {
  try {
    console.log('\n🔗 Connecting to database...');
    await connectDatabase();
  } catch (error) {
    console.error('Could not connect to database', { error });
    throw error;
  }

  const server = createServer(app.callback());

  server.listen(GRAPHQL_PORT, () => {
    console.info(`\n🚀 Server started at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}`);

    if (process.env.NODE_ENV !== 'production') {
      console.info(`\n🎠 GraphQL Playground available at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}/playground\n`);
    }

    SubscriptionServer.create(
      {
        onConnect: async (connectionParams: ConnectionParams) => {
          console.info('Client subscription connected', connectionParams);
          // @TODO - get user from connectionParams?.authorization
          // @TODO - return context with user
        },
        onDisconnect: () => console.info('Client subscription disconnected'),
        execute,
        subscribe,
        schema,
      },
      {
        server,
        path: '/subscriptions',
      },
    );
  });
};

(async () => {
  console.log('\n📡 Server starting...');
  await runServer();
})();
