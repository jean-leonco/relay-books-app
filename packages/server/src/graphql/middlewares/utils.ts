import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
