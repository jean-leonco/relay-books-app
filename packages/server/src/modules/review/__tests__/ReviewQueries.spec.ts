import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import {
  sanitizeTestObject,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
  gql,
} from '@workspace/test-utils';

import { createBook, createReview, createUser, getContext } from '../../../test/utils';

import schema from '../../../schema/schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('ReviewQueries', () => {
  it('should get review from node interface', async () => {
    const user = await createUser();
    const review = await createReview();

    const query = gql`
      query Q($id: ID!) {
        review: node(id: $id) {
          ... on Review {
            id
            rating
            user {
              name
            }
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('Review', review._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.review).not.toBe(null);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should not get review without user', async () => {
    const review = await createReview();

    const query = gql`
      query Q($id: ID!) {
        review: node(id: $id) {
          ... on Review {
            rating
            user {
              name
            }
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext();
    const variables = {
      id: toGlobalId('Review', review._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.review).toBe(null);
  });

  it('should get null review if isActive is false', async () => {
    const user = await createUser();
    const event = await createReview({ isActive: false });

    const query = gql`
      query Q($id: ID!) {
        review: node(id: $id) {
          ... on Review {
            rating
            user {
              name
            }
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('Review', event._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.review).toBe(null);
  });

  it('should query a connection of reviews', async () => {
    const user = await createUser();

    for (let i = 0; i < 6; i++) {
      await createReview();
    }

    const query = gql`
      query Q {
        reviews(first: 5) {
          edges {
            node {
              id
              rating
              user {
                name
              }
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.reviews).not.toBe(null);
    expect(result.data?.reviews.edges.length).toBe(5);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a null connection of reviews with there is no user on ctx', async () => {
    for (let i = 0; i < 6; i++) {
      await createReview();
    }

    const query = gql`
      query Q {
        reviews(first: 5) {
          edges {
            node {
              id
              rating
              user {
                name
              }
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext();
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.reviews).not.toBe(null);
    expect(result.data?.reviews.edges.length).toBe(0);
  });

  it('should query a connection of reviews with orderBy filter', async () => {
    const user = await createUser();

    const review1 = await createReview();
    const review2 = await createReview();
    const review3 = await createReview();

    const query = gql`
      query Q($filters: ReviewFilters) {
        reviews(filters: $filters) {
          edges {
            node {
              id
              rating
              user {
                name
              }
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        orderBy: { sort: 'CREATED_AT', direction: 'ASC' },
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.reviews).not.toBe(null);
    expect(result.data?.reviews.edges.length).toBe(3);
    expect(result.data?.reviews.edges[0].node.rating).toBe(review1.rating);
    expect(result.data?.reviews.edges[1].node.rating).toBe(review2.rating);
    expect(result.data?.reviews.edges[2].node.rating).toBe(review3.rating);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of reviews with book filter', async () => {
    const user = await createUser();

    const book = await createBook();
    for (let i = 0; i < 5; i++) {
      await createReview();
    }

    const book2 = await createBook();
    for (let i = 0; i < 3; i++) {
      await createReview({ bookId: book2._id });
    }

    const query = gql`
      query Q($filters: ReviewFilters) {
        reviews(filters: $filters) {
          edges {
            node {
              id
              rating
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        book: toGlobalId('Book', book._id),
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.reviews).not.toBe(null);
    expect(result.data?.reviews.edges.length).toBe(5);
    expect(result.data?.reviews.edges[0].node.book.name).toBe(book.name);
  });

  it('should query a connection of reviews with user filter', async () => {
    const user = await createUser();

    for (let i = 0; i < 3; i++) {
      await createReview();
    }

    const user2 = await createUser();
    for (let i = 0; i < 5; i++) {
      await createReview({ userId: user2._id });
    }

    const query = gql`
      query Q($filters: ReviewFilters) {
        reviews(filters: $filters) {
          edges {
            node {
              id
              rating
              user {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        user: toGlobalId('User', user._id),
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.reviews).not.toBe(null);
    expect(result.data?.reviews.edges.length).toBe(3);
    expect(result.data?.reviews.edges[0].node.user.name).toBe(user.name);
  });

  it('should query only active reviews', async () => {
    const user = await createUser();

    for (let i = 0; i < 5; i++) {
      await createReview();
    }

    for (let i = 0; i < 10; i++) {
      await createReview({ isActive: false });
    }

    const query = gql`
      query Q {
        reviews {
          edges {
            node {
              id
              rating
              user {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.reviews).not.toBe(null);
    expect(result.data?.reviews.edges.length).toBe(5);
  });
});
