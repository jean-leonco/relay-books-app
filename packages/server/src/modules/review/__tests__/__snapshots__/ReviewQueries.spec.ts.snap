// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review queries should get review from node interface 1`] = `
Object {
  "review": Object {
    "book": Object {
      "name": "name 1",
    },
    "id": "FROZEN-ID",
    "rating": 1,
    "user": Object {
      "name": "User #1",
    },
  },
}
`;

exports[`Review queries should query a connection of reviews 1`] = `
Object {
  "reviews": Object {
    "edges": Array [
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 6,
          "user": Object {
            "name": "User #1",
          },
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 5,
          "user": Object {
            "name": "User #1",
          },
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 4,
          "user": Object {
            "name": "User #1",
          },
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 3,
          "user": Object {
            "name": "User #1",
          },
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 2,
          "user": Object {
            "name": "User #1",
          },
        },
      },
    ],
  },
}
`;

exports[`Review queries should query a connection of reviews with orderBy filter 1`] = `
Object {
  "reviews": Object {
    "edges": Array [
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 1,
          "user": Object {
            "name": "User #1",
          },
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 2,
          "user": Object {
            "name": "User #1",
          },
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "rating": 3,
          "user": Object {
            "name": "User #1",
          },
        },
      },
    ],
  },
}
`;
